planner/core/rule_build_proj_below_agg.go

```diff
@@ -0,0 +1,168 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"fmt"
+
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/expression/aggregation"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+)
+
+// buildProjBelowAgg builds a ProjOperator below AggOperator.
+// If all the args of `aggFuncs`, and all the item of `groupByItems`
+// are columns or constants, we do not need to build the `proj`.
+func buildProjBelowAgg(plan PhysicalPlan) PhysicalPlan {
+	for _, child := range plan.Children() {
+		buildProjBelowAgg(child)
+	}
+
+	var aggFuncs []*aggregation.AggFuncDesc
+	var groupByItems []expression.Expression
+	if aggHash, ok := plan.(*PhysicalHashAgg); ok {
+		aggFuncs = aggHash.AggFuncs
+		groupByItems = aggHash.GroupByItems
+	} else if aggStream, ok := plan.(*PhysicalStreamAgg); ok {
+		aggFuncs = aggStream.AggFuncs
+		groupByItems = aggStream.GroupByItems
+	} else {
+		return plan
+	}
+
+	return doBuildProjBelowAgg(plan, aggFuncs, groupByItems)
+}
+
+func doBuildProjBelowAgg(aggPlan PhysicalPlan, aggFuncs []*aggregation.AggFuncDesc, groupByItems []expression.Expression) PhysicalPlan {
+	hasScalarFunc := false
+
+	// If the mode is FinalMode, we do not need to wrap cast upon the args,
+	// since the types of the args are already the expected.
+	if len(aggFuncs) > 0 && aggFuncs[0].Mode != aggregation.FinalMode {
```

Comment:
if the mode is `Partial2Mode`, we don't need the cast either? @XuHuaiyu @qw4990

Reply:
@eurekaka Seems you're right.

Reply:
Should I push another PR to fix this problem?  🤔

Reply:
ok